framework/src/convergence/DefaultMultiAppFixedPointConvergence.C

```diff
@@ -0,0 +1,214 @@
+  if (!_has_fixed_point_norm && parameters.isParamSetByUser("fixed_point_rel_tol"))
+    paramWarning(
+        "disable_fixed_point_residual_norm_check",
+        "fixed_point_rel_tol will be ignored because the fixed point residual check is disabled.");
+  if (!_has_fixed_point_norm && parameters.isParamSetByUser("fixed_point_abs_tol"))
+    paramWarning(
+        "disable_fixed_point_residual_norm_check",
+        "fixed_point_abs_tol will be ignored because the fixed point residual check is disabled.");
+  if (!_has_fixed_point_norm && parameters.isParamSetByUser("fixed_point_force_norms"))
+    paramWarning("disable_fixed_point_residual_norm_check",
+                 "fixed_point_force_norms will be ignored because the fixed point residual check "
+                 "is disabled.");
+
+  if (dynamic_cast<SteffensenSolve *>(&_fp_solve) && _fe_problem.hasMultiApps())
+  {
+    // Steffensen method uses half-steps
+    if (!parameters.isParamSetByAddParam("fixed_point_min_its"))
+      _min_fixed_point_its *= 2;
+    _max_fixed_point_its *= 2;
+  }
+}
+
+void
+DefaultMultiAppFixedPointConvergence::initialize()
+{
+  DefaultConvergenceBase::initialize();
+
+  _fixed_point_timestep_begin_norm.clear();
+  _fixed_point_timestep_end_norm.clear();
+  _fixed_point_timestep_begin_norm.resize(_max_fixed_point_its);
+  _fixed_point_timestep_end_norm.resize(_max_fixed_point_its);
+
+  _pp_history.str("");
+
+  // compute residual norm before iteration loop
+  if (_has_fixed_point_norm)
+  {
+    _fixed_point_initial_norm = _fe_problem.computeResidualL2Norm();
+    _console << COLOR_MAGENTA << "Initial fixed point residual norm: " << COLOR_DEFAULT;
+    if (_fixed_point_initial_norm == std::numeric_limits<Real>::max())
+      _console << " MAX ";
+    else
+      _console << std::scientific << _fixed_point_initial_norm;
+    _console << COLOR_DEFAULT << "\n" << std::endl;
+  }
+}
+
+void
+DefaultMultiAppFixedPointConvergence::preExecute()
+{
+  DefaultConvergenceBase::preExecute();
+
+  // compute TIMESTEP_BEGIN residual norm
+  if (_has_fixed_point_norm &&
+      (_fe_problem.hasMultiApps(EXEC_TIMESTEP_BEGIN) || _fixed_point_force_norms))
+  {
+    const auto iter = _fp_solve.numFixedPointIts() - 1;
+    _fixed_point_timestep_begin_norm[iter] = _fe_problem.computeResidualL2Norm();
+
+    Real begin_norm_old =
+        (iter > 0 ? _fixed_point_timestep_begin_norm[iter - 1] : std::numeric_limits<Real>::max());
+
+    _console << COLOR_MAGENTA << "Fixed point residual norm after TIMESTEP_BEGIN MultiApps: "
+             << Console::outputNorm(begin_norm_old, _fixed_point_timestep_begin_norm[iter])
+             << std::endl;
+  }
+}
+
+Convergence::MooseConvergenceStatus
+DefaultMultiAppFixedPointConvergence::checkConvergence(unsigned int iter)
+{
+  TIME_SECTION(_perfid_check_convergence);
+
+  if (_fixed_point_custom_pp)
+    computeCustomConvergencePostprocessor(iter);
+
+  // compute TIMESTEP_END residual norm
```

Comment:
same comments as preExecute()

Reply:
maybe checkConvergence should take an execute_on flag? what do you think>

Reply:
because there is some execute_on specific code here. So either we impose the execute_on to be correct, or we ask for it as a parameter